File: libs/power-log-parser/src/lib/parsers/start-of-game-trigger-parser.ts
import { BlockType, CardIds, GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Action, Node, NodeType } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

const FORCE_START_OF_GAME_POWERS: string[] = [CardIds.PrinceRenathal, CardIds.PrinceRenathal_CORE_REV_018];

export class StartOfGameTriggerParser implements ActionParser {
	readonly ParserName = 'StartOfGameTriggerParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		let action: Action | null = null;
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.Action &&
			(action = node.Object as Action).Type === (BlockType.TRIGGER as number) &&
			(action.TriggerKeyword === (GameTag.START_OF_GAME_KEYWORD as number) ||
				action.TriggerKeyword === (GameTag.TAG_NOT_SET as number))
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const action = node.Object as Action;
		const actionEntity = this.GameState.CurrentEntities.get(action.Entity);
		if (
			action.TriggerKeyword === (GameTag.TAG_NOT_SET as number) &&
			!FORCE_START_OF_GAME_POWERS.includes(actionEntity?.CardId ?? '')
		) {
			return null;
		}

		const controllerId = actionEntity?.GetTag(GameTag.CONTROLLER) ?? -1;
		return [
			GameEventProvider.Create(
				action.TimeStamp,
				'START_OF_GAME',
				GameEventHelper.CreateProvider(
					'START_OF_GAME',
					actionEntity?.CardId ?? '',
					controllerId,
					action.Entity,
					this.StateFacade,
				),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/StartOfGameTriggerParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using System.Linq;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

namespace HearthstoneReplays.Events.Parsers
{
    public class StartOfGameTriggerParser : ActionParser
    {
        private static List<string> FORCE_START_OF_GAME_POWERS = new List<string>() { CardIds.PrinceRenathal, CardIds.PrinceRenathal_CORE_REV_018 };

        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public StartOfGameTriggerParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            Action action = null;
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(Action)
                && (action = node.Object as Action).Type == (int)BlockType.TRIGGER
                && (action.TriggerKeyword == (int)GameTag.START_OF_GAME_KEYWORD || action.TriggerKeyword == (int)GameTag.TAG_NOT_SET);
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var action = node.Object as Action;
            var actionEntity = GameState.CurrentEntities.GetValueOrDefault(action.Entity);
            if (action.TriggerKeyword == (int)GameTag.TAG_NOT_SET && !FORCE_START_OF_GAME_POWERS.Contains(actionEntity?.CardId))
            {
                return null;
            }

            var controllerId = actionEntity?.GetTag(GameTag.CONTROLLER) ?? -1;
            return new List<GameEventProvider> { GameEventProvider.Create(
                action.TimeStamp,
                "START_OF_GAME",
                GameEvent.CreateProvider(
                    "START_OF_GAME",
                    actionEntity?.CardId,
                    controllerId,
                    action.Entity,
                    StateFacade,
                    null),
                true,
                node) 
            };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
